### PR TITLE
Mod inverse focus problem

### PR DESCRIPTION
**Pull request by andrewwangva**
I felt like the inverse section needed a focus problem and I also cleared up some confusing portions.

### CODE REVIEW

File: content/4_Gold/Modular.mdx [hunk 1]
```diff
@@ -263,6 +266,65 @@ public class main
...
+    factorial[0] = factorial[1] =1;
+    for (int i = 2; i<=1000000; i++) {
+        factorial[i]=(factorial[i-1]*i)%MOD;
+    }
+ 
+    ll n, a, b;
+    cin >> n;
+    for(int i = 0; i < n; i++) {
+        cin >> a >> b;
+        ll ans = (factorial[a] * inverse(factorial[b])) % MOD;
```
Comment (bqi343) [inline content/4_Gold/Modular.mdx:315]: worth noting that if you compute `inverse(factorial[n])` then you can get `inverse(factorial[1..n-1])` in O(n) time.

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/4_Gold/Modular.mdx [hunk 2]
```diff
@@ -181,42 +181,45 @@ public class Exponentation
...
-(2x)\cdot i\equiv x\cdot (2i)\equiv x\pmod{10^9+7}.
+1\equiv p+1 \pmod{p}.
 $$
 
-For example, $10i\equiv 5\pmod{10^9+7}$.
-
-<!-- #### Example: Sprinkers 2
+$$
+2\cdot i\equiv p+1 \pmod{p}.
+$$
```
Comment (frodakcin) [inline content/4_Gold/Modular.mdx:208]: $i$ kinda comes out of nowhere. Could be beneficial to either explain it beforehand, or just use $2^{-1}$ as notation?

**Review by frodakcin — APPROVED** [review]
Fine as is. Possible changes below

* More explicitly define $i$/use alternate notation?
* Could also briefly mention totient function theorem but it might just make module more confusing

**Comment by andrewwangva** [comment]
> Fine as is. Possible changes below
> 
> * More explicitly define $i$/use alternate notation?
> * Could also briefly mention totient function theorem but it might just make module more confusing

I think adding totient function isn't necessary since the resource talks abt it + fermat's is kinda an extension on it anyways

File: content/4_Gold/Modular.mdx [hunk 3]
```diff
@@ -328,7 +328,7 @@ int main() {
 Because it takes $\mathcal{O}(\log p)$ time to compute a modular inverse modulo
 $p$, frequent use of division inside a loop can significantly increase the
 running time of a program. If the modular inverse of the same number(s) is/are
-being used many times, it is a good idea to precalculate it.
+being used many times, it is a good idea to precalculate it. Note the above problem could also be solved by precalculating invererses by first calculating the inverse of n! for a large n then finding all previous values by multiplying in $O(n)$.
```
Comment (bqi343) [inline content/4_Gold/Modular.mdx:331]: would like to see this in the code. 

+ typos

+ n! and n should be surrounded by $$

File: content/4_Gold/Modular.mdx [hunk 4]
```diff
@@ -198,7 +198,7 @@ The **modular inverse** is the equivalent of the reciprocal in real-number
 arithmetic; to divide $a$ by $b$, multiply $a$ by the modular inverse of $b$.
 We'll only consider prime moduli $p$ here.
 
-For example, the inverse of $2 \pmod{p}$ where $p=10^9+7$ is
+For example, $i$, the modular inverse of $2$ where $p=10^9+7$ is
```
Comment (bqi343) [inline content/4_Gold/Modular.mdx:201]: now you say that $i$ is the modular inverse twice

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/4_Gold/Modular.mdx (hunk 3 above)
Comment (andrewwangva) [reply content/4_Gold/Modular.mdx:331]: as a comment?

Comment (bqi343) [reply content/4_Gold/Modular.mdx:331]: actual code

Comment (dongliu0426) [reply content/4_Gold/Modular.mdx:331]: add \mathcal too

**Comment by bqi343** [comment]
@andrewwangva update?

**Comment by andrewwangva** [comment]
> @andrewwangva update?

lmao i forgot abt this - i'll finish this in the next 3 days hopefully.

File: content/4_Gold/Modular.mdx [hunk 5]
```diff
@@ -263,15 +266,118 @@ public class main
...
+
+</LanguageSection>
+
+### Solution - Binomial Coefficients $\mathcal{O}(N)$
+
 Because it takes $\mathcal{O}(\log p)$ time to compute a modular inverse modulo
 $p$, frequent use of division inside a loop can significantly increase the
 running time of a program. If the modular inverse of the same number(s) is/are
-being used many times, it is a good idea to precalculate it.
+being used many times, it is a good idea to precalculate it. Note the above problem could also be solved by precalculating invererses. First, find the inverse of $N!$ for a large $N$, then calculate the inverses of $1...{N!-1}$ by multiplying each inverse starting from $N!$ with their respective values in $\mathcal{O}(N)$.
```
Comment (bqi343) [inline content/4_Gold/Modular.mdx:332]: > invererses

> N!-1

> multiplying each inverse starting from $N!$ with their respective values

what does this mean?

File: content/4_Gold/Modular.mdx [hunk 6]
```diff
@@ -263,15 +266,118 @@ public class main
...
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+ 
+using ll = long long;
+const int MOD = (int)1e9+7;
+ 
+ll inverse(ll num){
```
Comment (bqi343) [inline content/4_Gold/Modular.mdx:290]: indent with tabs, same below

File: content/4_Gold/Modular.mdx [hunk 7]
```diff
@@ -181,42 +181,45 @@ public class Exponentation
...
+2\cdot i\equiv p+1 \pmod{p}.
+$$
 
-There aren't many good examples of Modular Inverses in USACO. One problem where it was useful was the platinum problem "Sprinkers 2".
+$$
+i=\frac{p+1}{2}=5\cdot 10^8+4
+$$
 
-<FocusProblem problem="dp[2]" />
+where $i$ is the modular inverse. This means that for any integer $x$,
```
Comment (bqi343) [inline content/4_Gold/Modular.mdx:214]: still duplicated

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/4_Gold/Modular.mdx [hunk 8]
```diff
@@ -263,15 +265,125 @@ public class main
 
 </LanguageSection>
 
+### Solution - Binomial Coefficients $\mathcal{O}(N \log N)$
```
Comment (bqi343) [inline content/4_Gold/Modular.mdx:268]: where is the $\log N$ coming from? shouldn't it be $\log MOD$?

File: content/4_Gold/Modular.mdx [hunk 9]
```diff
@@ -263,15 +265,125 @@ public class main
...
+	}
+	return res;
+}
+
+int main() {
+	cin.tie(0)->sync_with_stdio(0);
+	ll factorial[1000001];
+	factorial[0] = factorial[1] =1;
+	for (int i = 2; i<=1000000; i++) {
+		factorial[i]=(factorial[i-1]*i)%MOD;
```
Comment (bqi343) [inline content/4_Gold/Modular.mdx:309]: inconsistent spacing around `=` and `<=`

File: content/4_Gold/Modular.mdx [hunk 10]
```diff
@@ -263,15 +265,125 @@ public class main
...
+
+### Solution - Binomial Coefficients $\mathcal{O}(N)$
+
 Because it takes $\mathcal{O}(\log p)$ time to compute a modular inverse modulo
 $p$, frequent use of division inside a loop can significantly increase the
 running time of a program. If the modular inverse of the same number(s) is/are
-being used many times, it is a good idea to precalculate it.
+being used many times, it is a good idea to precalculate it. 
+
+To precalculate the inverses of $1...{(N-1)!}$, first, find the
```
Comment (bqi343) [inline content/4_Gold/Modular.mdx:335]:
```suggestion
To precalculate the inverses of $1!,\,2!,\, \ldots,\, N!$, first find the
inverse of $N!$. Then calculate the inverses of $(N-1)!, \,(N-2)!,\,\dots,\,1!$ by
multiplying each inverse starting from $N!$ with their respective values in
$\mathcal{O}(N)$.
```

Comment (bqi343) [reply content/4_Gold/Modular.mdx:335]: what are the braces for?

File: content/4_Gold/Modular.mdx [hunk 11]
```diff
@@ -263,15 +265,125 @@ public class main
...
+inverse of $N!$ for a large $N$. Then calculate the inverses of ${(N-1)!}, {(N-2)!},...1$ by
+multiplying each inverse starting from $N!$ with their respective values in
```
Comment (bqi343) [inline content/4_Gold/Modular.mdx:337]: it's not clear what "respective values" means

File: content/4_Gold/Modular.mdx [hunk 12]
```diff
@@ -263,15 +265,125 @@ public class main
...
+$\mathcal{O}(N)$.
 
 Also, one must always ensure that they do not attempt to divide by 0. Be aware
```
Comment (bqi343) [inline content/4_Gold/Modular.mdx:340]: but this isn't an issue for binomial coefficients, so why is this located here?

File: content/4_Gold/Modular.mdx [hunk 13]
```diff
@@ -263,15 +265,125 @@ public class main
...
+	factorial[0] = factorial[1] =1;
+	for (int i = 2; i<=1000000; i++) {
+		factorial[i]=(factorial[i-1]*i)%MOD;
+	}
+
+	ll n, a, b;
+	cin >> n;
+	for(int i = 0; i < n; i++) {
+		cin >> a >> b;
+		ll ans = (factorial[a] * inverse(factorial[b])) % MOD;
```
Comment (bqi343) [inline content/4_Gold/Modular.mdx:316]:
```suggestion
		ll ans = factorial[a] * inverse(factorial[b]) % MOD;
```

File: content/4_Gold/Modular.mdx [hunk 14]
```diff
@@ -263,15 +265,125 @@ public class main
...
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+using ll = long long;
+const int MOD = (int)1e9+7;
+
+ll inverse(ll num){
+	ll a = num; ll b = MOD-2;
```
Comment (bqi343) [inline content/4_Gold/Modular.mdx:292]: why not replace `num` with `a` on line 291?

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**